Refactor Proxy Server configuration and Formatting

evan-m-jackson opened the pull request:
- Rename classes/files to have more appropriate names and made sure test following similar naming conventions to classes  they are testing
-  Update YAML files for testing
-  Refactor Proxy Server classes to make them more efficient and less coupled
-  Fix spacing and other formatting issues

evan-m-jackson committed:
Refactor Proxy Server Configuration and Formatting

- Rename classes/files to have more appropriate names and made sure test following similar naming conventions to classes  they are testing
- Separate code/tests into properly named folders
- Add new method 'GetCode' to Header class
- Add new 'Port' class allow server to run on any valid port given by the user
- Add additional status code items to 'ResponseCode' classes dictionary variable
- Add new class, 'UpdateInitialLine', to refactor the path name to the first available id in the Todo API and/or refactor the path if a DELETE request is being made
- Add new class, 'FirstID', to return first available id in the Todo API, as a string
- Add new class, 'CheckPath', to return a boolean to see if the 'todo' path is valid

evan-m-jackson committed:
Refactor YAML files and Acceptance Tests

- Rename Tar building step in Azure deploy yaml file
- Refactor 'Rakefile' to have separate tests for GitHub Actions called ':acctest' and added methods to return an exit code if the acceptance tests have errors/failures.
- Add 'Acceptance Test' step to dotnet.yml in order to run acceptance tests from 'http-server-spec'
- Refactor README file to include instructions on running acceptance tests and running the Todo API Django Server.

evan-m-jackson committed:
Refactor Code and Add Testing

- Remove empty interface, 'IProxyClient', and unnecessary test double 'TestProxyClient'
- Add unit tests for 'Close' and 'Flush' methods in 'IWriteStreams'
- Add unit test for 'Close' method in 'IReadStreams'

evan-m-jackson committed:
Update TestYAML file

skp96 commented on HTTPServerProject.Tests/HTTPServerProxyTests/CheckPathTests.cs:
```diff
@@ -0,0 +1,16 @@
+public class HttpPathTests
+{
+    [Theory]
+    [InlineData("todo", true)]
```
`InlineData` is cool! Nice job reducing the number of tests you would have to write.

evan-m-jackson replied:
Thank you!

skp96 commented on HTTPServerProject.Tests/HTTPServerProxyTests/CheckPathTests.cs:
```diff
@@ -0,0 +1,16 @@
+    [Theory]
+    [InlineData("todo", true)]
+    [InlineData("tod", false)]
+    public void GetTodoPathTrueTest(string path, bool expected)
```
The name of the test suggests you're testing for a `true` scenario, but I think this test would run twice; (1) testing a scenario where you can find the path `todo` (e.g. `true`) and (2) testing a scenario where you cannot find the path `todo` (e.g. `false`). Maybe the test name should be changed to `ValidateTodoPathTest`? What are your thoughts?

evan-m-jackson replied:
That's a good point, I will update for that!

paytonrules commented:
Can you add a description to the PR? Especially with all these commits squashed it's extremely helpful for readers. Many clients will even have a template you have to follow.

paytonrules commented on HTTPServerProject.Tests/HTTPServerProxyTests/ResponseTests.cs:
```diff
@@ -0,0 +1,224 @@
+public class ProxyResponseTests
+{
+    [Fact]
+    public void GetTodoPOSTValidResponseTest()
```
Why do all these start with `GetTodo`? Is this a name that hung around from previous versions?

evan-m-jackson replied:
No, it's just a silly name. It probably would be better to swap out `GetTodo` for `Proxy`

paytonrules commented on HTTPServerProject.Tests/HTTPServerProxyTests/ResponseTests.cs:
```diff
@@ -0,0 +1,224 @@
+        var expectedStatusLine = "HTTP/1.1 201 Created";
+        var proxyStream = new List<string>() 
+		{
+        "HTTP/1.1 200 OK", 
```
Keep an eye on these formatting issues.

evan-m-jackson replied:
Will do. The formatting doesn't look that bad in Ryder 😂

paytonrules commented on HTTPServerProject.Tests/HTTPServerProxyTests/ResponseTests.cs:
```diff
@@ -0,0 +1,224 @@
+		"", 
+		"{ 'task': 'First Item' }" 
+		};
+        var stream = new List<string>();
```
It's a small thing - but you should put stuff that depends on each other together. So you create the stream on line 21 but inject it on 25. You the POST and todo crated on 23 and 24 then you hop over that to use them on 27 (with the writer created in the middle). 

If you put it like:

```
var proxyStream = blah;
var reader = new TestReadStreams(proxyStream);

var stream = new List<string>();
var writer = new TestWriteStreams(stream);

var httpType = "POST";
var httpPath = "todo";
var proxyResponse = new ProxyResponse....
```

Now it's a little easier to read at a glance.

evan-m-jackson replied:
Agreed.

paytonrules commented on HTTPServerProject.Tests/HTTPServerProxyTests/ResponseTests.cs:
```diff
@@ -0,0 +1,224 @@
+        var proxyResponse = new ProxyResponse(reader: reader, writer: writer, path: httpPath, type: httpType);
+        proxyResponse.GetResponse();
+        var actualStatusLine = stream[0];
+        Assert.Equal(expectedStatusLine, actualStatusLine);
```
Moving expectedStatusLine down here as well is also a bit easier to read.

evan-m-jackson replied:
Agreed.

paytonrules commented on HTTPServerProject.Tests/HTTPServerProxyTests/ResponseTests.cs:
```diff
@@ -0,0 +1,224 @@
+        var writer = new TestWriteStreams(stream);
+
+        var proxyResponse = new ProxyResponse(reader: reader, writer: writer, path: httpPath, type: httpType);
+        proxyResponse.GetResponse();
```
The name `GetResponse` doesn't make sense. It appears to write to the write stream, it doesn't return a response. Are you trying to use the command pattern here? (quick refresher) https://www.tutorialspoint.com/design_pattern/command_pattern.htm

evan-m-jackson replied:
Yes, I was trying to do something like that. I can rename the `GetResponse` method to `Execute`

paytonrules commented on HTTPServerProject/HTTPServerProxy/Response.cs:
```diff
@@ -0,0 +1,135 @@
+            var response = new WriteResponse(_writer, status);
+            response.GetResponse();
+        }
+        else
```
You can shrink this up with an `else if (code == 200)`

evan-m-jackson replied:
Agreed.

paytonrules commented on HTTPServerProject/HTTPServerProxy/Response.cs:
```diff
@@ -0,0 +1,135 @@
+        _type = type;
+    }
+
+    public void GetResponse()
```
I mentioned this in the tests - but `GetResponse` implies this returns a response so it should have another name.

evan-m-jackson replied:
Agreed.

paytonrules commented on HTTPServerProject.Tests/HTTPServerProxyTests/TestFirstID.cs:
```diff
@@ -0,0 +1,103 @@
+
+namespace HTTPServerProxyTests.FirstID;
+
+public class TestGetFirstID : IGetFirstID
```
I think I left a comment about this on the other PR - so this might be redundant - but this class looks like a duplicate of a real class that belongs in the implementation. If it's a mock, it could it be made a lot simpler.

evan-m-jackson replied:
Yes it is a mock - I am mocking a method that makes a GET request. I could take this method out and give the class less responsibility.

paytonrules commented on HTTPServerProject.Tests/HTTPServerProxyTests/UpdateInitialLineTests.cs:
```diff
@@ -0,0 +1,26 @@
+using HTTPServerReadTests.Streams;
+using HTTPServerWriteTests.Streams;
+
+namespace HTTPServerProject.Tests.Update.Initial.Line;
```
Shouldn't this namespace name end at HTTPServerProject.Tests ?

evan-m-jackson replied:
Agreed. I might use this namespace for all the tests

paytonrules commented on HTTPServerProject.Tests/HTTPServerReadTests/TestReadStreams.cs:
```diff
@@ -0,0 +1,104 @@
+namespace HTTPServerReadTests.Streams;
+
+public class UnitTestsForReadStreams
+    {
```
More weird formatting going on here.

paytonrules commented on HTTPServerProject.Tests/HTTPServerReadTests/TestReadStreams.cs:
```diff
@@ -0,0 +1,104 @@
+
+namespace HTTPServerReadTests.Streams;
+
+public class UnitTestsForReadStreams
```
This looks like a test for `TestReadStreams` - you don't need to test test doubles, you can just use them in your tests.

evan-m-jackson replied:
Sounds good.

paytonrules commented on HTTPServerProject.Tests/HTTPServerReadTests/TestReadStreams.cs:
```diff
@@ -0,0 +1,104 @@
+
+    public void Close()
+    {
+		closeCalled = true;
```
That's a lot of spaces

evan-m-jackson replied:
Agreed.

paytonrules commented on HTTPServerProject.Tests/HTTPServerReadTests/TestReadStreams.cs:
```diff
@@ -0,0 +1,104 @@
+            }
+            else
+            {
+                sArr[0] = "";
```
What if the input is only one character, won't you lose the character?

evan-m-jackson replied:
No, because the one character would assigned to `var c` before the `if/else` statement.

paytonrules commented on HTTPServerProject.Tests/HTTPServerResponseTests/ByPathTests.cs:
```diff
@@ -0,0 +1,365 @@
+        var pathDict = pathParams.pathDict;
+
+        var responsePath = new ResponsePath(writer, pathDict);
+        responsePath.ExecuteRequest(path: path, type: type, requestBody: requestBody);
```
Would probably name this just `Execute` because Request has multiple meanings here.

evan-m-jackson replied:
Agreed.

paytonrules commented on HTTPServerProject.Tests/HTTPServerResponseTests/TestParameters.cs:
```diff
@@ -0,0 +1,95 @@
+
+public class TestPathParameters : IPathParameters
+{
+    public Dictionary<string, Dictionary<string, Dictionary<string, object>>> pathDict = new Dictionary<string, Dictionary<string, Dictionary<string, object>>>();
```
You can use `using` to make this type way shorter, for example `using MyAlias = System.Collections.Generic.List<System.Tuple<int, string, int>>;`

evan-m-jackson replied:
I couldn't get this to work for this constructor variable.

paytonrules commented on HTTPServerProject.Tests/HTTPServerWriteTests/TestWriteStreams.cs:
```diff
@@ -0,0 +1,64 @@
+using HTTPServerWrite.Streams;
+
+namespace HTTPServerWriteTests.Streams
```
This is inconsistent with the way you've been defining namespaces everywhere else.

evan-m-jackson replied:
Will update.

paytonrules commented on HTTPServerProject.Tests/HTTPServerWriteTests/TestWriteStreams.cs:
```diff
@@ -0,0 +1,64 @@
+    public class UnitTestsForWriteStreams
+    {
+        [Fact]
+        public void WriterIsClosedTest()
```
What about when you don't call `Close` - it shouldn't have called `close` - correct?

paytonrules replied:
Oh but this is another test for a dummy implementation. You generally don't need those.

evan-m-jackson replied:
Sounds good.

paytonrules commented on HTTPServerProject.Tests/HTTPServerWriteTests/WriteRequestTests.cs:
```diff
@@ -0,0 +1,85 @@
+
+        Assert.Equal(expected, stream);
+    }
+}
```
I don't understand - what do these tests have to do with the proxy?

evan-m-jackson replied:
I updated the naming of the test. I also removed the `GetProxyResponseTest` because it has nothing to do with the `WriteRequest` class

paytonrules commented on HTTPServerProject/HTTPServerProxy/Client.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    TcpClient client = new TcpClient("127.0.0.1", 8000);
+
+    public NetworkStream GetStream()
```
I'm not sure what purpose this class is serving anymore - it's just a getter. Now maybe if it TOLD the client to do something, I could see it.

evan-m-jackson replied:
Agreed, I'm going to remove this class.

skp96 commented on HTTPServerProject.Tests/HTTPServerProxyTests/ResponseTests.cs:
```diff
@@ -0,0 +1,224 @@
+    public void GetTodoPOSTUnsupportedMediaTest()
+    {
+        var expectedStatusLine = "HTTP/1.1 415 Unsupported Media Type";
+        var proxyStream = new List<string>() { "HTTP/1.1 415 Unsupported Media Type", "" };
+        var stream = new List<string>();
+        var reader = new TestReadStreams(proxyStream);
+        var httpType = "POST";
+        var httpPath = "todo";
+        var writer = new TestWriteStreams(stream);
+
+        var proxyResponse = new ProxyResponse(reader: reader, writer: writer, path: httpPath, type: httpType);
+        proxyResponse.GetResponse();
```
Noticing the setup for each test is the same or very similar. Can you introduce a test fixture to remove some of the duplications?

evan-m-jackson replied:
Yes, I was able to use `MemberData` to remove duplicates.

evan-m-jackson committed:
Refactor Code and Formatting

- Refactor all test namespaces to be the same (HTTPServerProject.Tests)
- Refactor 'ExecuteRequest' to 'Execute' in 'ResponsePath' class
- Refactor 'ResponsePath' tests to use an 'IEnumerable' and remove duplication
- Delete 'Client' class and use 'TcpClient' class in 'Program'
- Refactor 'GetFirstID' class to not inherit from "IGetFirstID" (which was deleted) and remove methods that were giving the class too many responsibilities
- Create 'TodoList' class that has the methods that were removed from 'GetFirstID'
- Remove 'TestGetFirstID' class that was a test double for the 'GetFirstID' class and create unit test instead
- Refactor 'GetRequest' to 'Run' in 'WriteRequest' class
- Refactor 'GetResponse' to 'Run' in 'ProxyResponse' class and shortened if/else statement in the method
- Refactor 'GetResponse' to 'Run' in 'WriteResponse' class
- Refactor 'ProxyResponse' tests to use an 'IEnumerable' and remove duplication
- Remove unit tests for 'TestReadStreams' and 'TestWriteStreams' classes

evan-m-jackson committed:
Refactor spacing

evan-m-jackson committed:
Refactor 'PathParameters' class public variable 'pathDict'

paytonrules reviewed (approved):
Can the other, older PR go away now?
